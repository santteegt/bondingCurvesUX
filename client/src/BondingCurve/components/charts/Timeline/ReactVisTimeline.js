
import React, { PureComponent } from 'react'
import moment from 'moment'
import PropTypes from 'prop-types'
import { AreaSeries, FlexibleWidthXYPlot, GradientDefs, LineSeries, MarkSeries, VerticalGridLines, XAxis } from 'react-vis'
import styles from './timeline.module.scss'

export default class ReactVisTimeline extends PureComponent {
    state = {
        hoverValues: null
    }

    static propTypes = {
        onShowDetail: PropTypes.func,
        minDomain: PropTypes.number,
        height: PropTypes.number.isRequired,
        data: PropTypes.array.isRequired
    }

    _onMouseLeave = () => {
        const { onShowDetail } = this.props

        this.setState({ hoverValues: null })

        if (onShowDetail) {
            onShowDetail()
        }
    };

    _onNearestX = (value) => {
        const { onShowDetail } = this.props

        this.setState({ hoverValues: [value] })

        if (onShowDetail) {
            onShowDetail(value)
        }
    };

    render() {
        const { hoverValues } = this.state
        const { minDomain, height, data } = this.props

        const domain = minDomain ? [minDomain, moment().endOf('day').valueOf()] : null

        return (
            <div className={styles.ocean_chart}>
                <FlexibleWidthXYPlot
                    margin={{ left: 0, right: 0, top: 10, bottom: 40 }}
                    xType="time"
                    animation
                    onMouseLeave={this._onMouseLeave}
                    height={height}
                    xDomain={domain}>

                    <LineSeries
                        strokeWidth={3}
                        className={styles.ocean_line}
                        style={{ strokeLinejoin: 'round' }}
                        onNearestX={this._onNearestX}
                        data={data}
                    />

                    <GradientDefs>
                        <linearGradient id="oceanGradient" x1="0%" y1="0%" x2="0%" y2="100%">
                            <stop offset="0%" stopColor="#ff4092" stopOpacity={0.8} />
                            <stop offset="90%" stopColor="white" stopOpacity={0} />
                        </linearGradient>
                    </GradientDefs>

                    <AreaSeries
                        color={'url(#oceanGradient)'}
                        data={data}
                    />

                    {
                        hoverValues && (
                            <VerticalGridLines
                                className={styles.ocean_crosshair_line}
                                tickValues={[hoverValues[0].x]}
                            />
                        )
                    }

                    {
                        hoverValues && (
                            <MarkSeries
                                className={styles.ocean_crosshair_dot}
                                size={4}
                                data={hoverValues}
                            />
                        )
                    }

                    {
                        hoverValues && (
                            <XAxis
                                tickTotal={6}
                            />
                        )
                    }

                </FlexibleWidthXYPlot>
            </div>
        )
    }
}
