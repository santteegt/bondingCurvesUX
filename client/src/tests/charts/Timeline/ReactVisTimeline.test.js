import React from 'react';
import ReactDOM from 'react-dom';
import ReactVisTimeline from '../../../BondingCurve/components/charts/Timeline/ReactVisTimeline';
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Crosshair, XAxis, MarkSeries, VerticalGridLines } from 'react-vis';

Enzyme.configure({ adapter: new Adapter() })

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<ReactVisTimeline height={200} data={[]} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('should match snapshot', () => {
  const wrapper = shallow(
    <ReactVisTimeline height={200} data={[]} />
  );
  expect(wrapper).toMatchSnapshot()
});

it('should show hover elements is state has hoverValues', () => {
  const wrapper = shallow(
    <ReactVisTimeline height={200} data={[]} />
  );
  wrapper.setState({ hoverValues: [{ x: 0, y: 0 }] });

  expect(
    wrapper.containsMatchingElement(
      <VerticalGridLines />
    )
  ).toBe(true);

  expect(
    wrapper.containsMatchingElement(
      <XAxis />
    )
  ).toBe(true);

  expect(
    wrapper.containsMatchingElement(
      <MarkSeries />
    )
  ).toBe(true);
});

it('should handle _onMouseLeave', () => {
  const onShowDetail = jest.fn();
  const wrapper = shallow(
    <ReactVisTimeline height={200} data={[]} onShowDetail={onShowDetail} />
  );

  wrapper.instance()._onMouseLeave();

  expect(onShowDetail.mock.calls.length).toBe(1);
  expect(onShowDetail).toBeCalledWith();

  expect(wrapper.state('hoverValues')).toBe(null)

});


it('should handle _onNearestX', () => {
  const onShowDetail = jest.fn();
  const wrapper = shallow(
    <ReactVisTimeline height={200} data={[]} onShowDetail={onShowDetail} />
  );
  const nearest = { x: "x", y: "y" };

  wrapper.instance()._onNearestX(nearest);

  expect(onShowDetail.mock.calls.length).toBe(1);
  expect(onShowDetail).toBeCalledWith(nearest);

  expect(wrapper.state('hoverValues')[0]).toBe(nearest)

});