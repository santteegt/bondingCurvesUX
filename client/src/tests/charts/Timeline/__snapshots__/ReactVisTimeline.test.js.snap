// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReactVisTimeline
    data={Array []}
    height={200}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <FlexibleXYPlot
        animation={true}
        height={200}
        margin={
          Object {
            "bottom": 40,
            "left": 0,
            "right": 0,
            "top": 10,
          }
        }
        onMouseLeave={[Function]}
        xDomain={null}
        xType="time"
      >
        <LineSeries
          className="ocean_line"
          curve={null}
          data={Array []}
          getNull={[Function]}
          onNearestX={[Function]}
          opacity={1}
          stack={false}
          strokeStyle="solid"
          strokeWidth={3}
          style={
            Object {
              "strokeLinejoin": "round",
            }
          }
        />
        <GradientDefs
          className=""
        >
          <linearGradient
            id="oceanGradient"
            x1="0%"
            x2="0%"
            y1="0%"
            y2="100%"
          >
            <stop
              offset="0%"
              stopColor="#ff4092"
              stopOpacity={0.8}
            />
            <stop
              offset="90%"
              stopColor="white"
              stopOpacity={0}
            />
          </linearGradient>
        </GradientDefs>
        <AreaSeries
          className=""
          color="url(#oceanGradient)"
          data={Array []}
          getNull={[Function]}
          stack={false}
          style={Object {}}
        />
      </FlexibleXYPlot>,
      "className": "ocean_chart",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "animation": true,
        "children": Array [
          <LineSeries
            className="ocean_line"
            curve={null}
            data={Array []}
            getNull={[Function]}
            onNearestX={[Function]}
            opacity={1}
            stack={false}
            strokeStyle="solid"
            strokeWidth={3}
            style={
              Object {
                "strokeLinejoin": "round",
              }
            }
          />,
          <GradientDefs
            className=""
          >
            <linearGradient
              id="oceanGradient"
              x1="0%"
              x2="0%"
              y1="0%"
              y2="100%"
            >
              <stop
                offset="0%"
                stopColor="#ff4092"
                stopOpacity={0.8}
              />
              <stop
                offset="90%"
                stopColor="white"
                stopOpacity={0}
              />
            </linearGradient>
          </GradientDefs>,
          <AreaSeries
            className=""
            color="url(#oceanGradient)"
            data={Array []}
            getNull={[Function]}
            stack={false}
            style={Object {}}
          />,
          null,
          null,
          null,
        ],
        "height": 200,
        "margin": Object {
          "bottom": 40,
          "left": 0,
          "right": 0,
          "top": 10,
        },
        "onMouseLeave": [Function],
        "xDomain": null,
        "xType": "time",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "ocean_line",
            "curve": null,
            "data": Array [],
            "getNull": [Function],
            "onNearestX": [Function],
            "opacity": 1,
            "stack": false,
            "strokeStyle": "solid",
            "strokeWidth": 3,
            "style": Object {
              "strokeLinejoin": "round",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <linearGradient
              id="oceanGradient"
              x1="0%"
              x2="0%"
              y1="0%"
              y2="100%"
            >
              <stop
                offset="0%"
                stopColor="#ff4092"
                stopOpacity={0.8}
              />
              <stop
                offset="90%"
                stopColor="white"
                stopOpacity={0}
              />
            </linearGradient>,
            "className": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <stop
                  offset="0%"
                  stopColor="#ff4092"
                  stopOpacity={0.8}
                />,
                <stop
                  offset="90%"
                  stopColor="white"
                  stopOpacity={0}
                />,
              ],
              "id": "oceanGradient",
              "x1": "0%",
              "x2": "0%",
              "y1": "0%",
              "y2": "100%",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "offset": "0%",
                  "stopColor": "#ff4092",
                  "stopOpacity": 0.8,
                },
                "ref": null,
                "rendered": null,
                "type": "stop",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "offset": "90%",
                  "stopColor": "white",
                  "stopOpacity": 0,
                },
                "ref": null,
                "rendered": null,
                "type": "stop",
              },
            ],
            "type": "linearGradient",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "",
            "color": "url(#oceanGradient)",
            "data": Array [],
            "getNull": [Function],
            "stack": false,
            "style": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        null,
        null,
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <FlexibleXYPlot
          animation={true}
          height={200}
          margin={
            Object {
              "bottom": 40,
              "left": 0,
              "right": 0,
              "top": 10,
            }
          }
          onMouseLeave={[Function]}
          xDomain={null}
          xType="time"
        >
          <LineSeries
            className="ocean_line"
            curve={null}
            data={Array []}
            getNull={[Function]}
            onNearestX={[Function]}
            opacity={1}
            stack={false}
            strokeStyle="solid"
            strokeWidth={3}
            style={
              Object {
                "strokeLinejoin": "round",
              }
            }
          />
          <GradientDefs
            className=""
          >
            <linearGradient
              id="oceanGradient"
              x1="0%"
              x2="0%"
              y1="0%"
              y2="100%"
            >
              <stop
                offset="0%"
                stopColor="#ff4092"
                stopOpacity={0.8}
              />
              <stop
                offset="90%"
                stopColor="white"
                stopOpacity={0}
              />
            </linearGradient>
          </GradientDefs>
          <AreaSeries
            className=""
            color="url(#oceanGradient)"
            data={Array []}
            getNull={[Function]}
            stack={false}
            style={Object {}}
          />
        </FlexibleXYPlot>,
        "className": "ocean_chart",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animation": true,
          "children": Array [
            <LineSeries
              className="ocean_line"
              curve={null}
              data={Array []}
              getNull={[Function]}
              onNearestX={[Function]}
              opacity={1}
              stack={false}
              strokeStyle="solid"
              strokeWidth={3}
              style={
                Object {
                  "strokeLinejoin": "round",
                }
              }
            />,
            <GradientDefs
              className=""
            >
              <linearGradient
                id="oceanGradient"
                x1="0%"
                x2="0%"
                y1="0%"
                y2="100%"
              >
                <stop
                  offset="0%"
                  stopColor="#ff4092"
                  stopOpacity={0.8}
                />
                <stop
                  offset="90%"
                  stopColor="white"
                  stopOpacity={0}
                />
              </linearGradient>
            </GradientDefs>,
            <AreaSeries
              className=""
              color="url(#oceanGradient)"
              data={Array []}
              getNull={[Function]}
              stack={false}
              style={Object {}}
            />,
            null,
            null,
            null,
          ],
          "height": 200,
          "margin": Object {
            "bottom": 40,
            "left": 0,
            "right": 0,
            "top": 10,
          },
          "onMouseLeave": [Function],
          "xDomain": null,
          "xType": "time",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "ocean_line",
              "curve": null,
              "data": Array [],
              "getNull": [Function],
              "onNearestX": [Function],
              "opacity": 1,
              "stack": false,
              "strokeStyle": "solid",
              "strokeWidth": 3,
              "style": Object {
                "strokeLinejoin": "round",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <linearGradient
                id="oceanGradient"
                x1="0%"
                x2="0%"
                y1="0%"
                y2="100%"
              >
                <stop
                  offset="0%"
                  stopColor="#ff4092"
                  stopOpacity={0.8}
                />
                <stop
                  offset="90%"
                  stopColor="white"
                  stopOpacity={0}
                />
              </linearGradient>,
              "className": "",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <stop
                    offset="0%"
                    stopColor="#ff4092"
                    stopOpacity={0.8}
                  />,
                  <stop
                    offset="90%"
                    stopColor="white"
                    stopOpacity={0}
                  />,
                ],
                "id": "oceanGradient",
                "x1": "0%",
                "x2": "0%",
                "y1": "0%",
                "y2": "100%",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "offset": "0%",
                    "stopColor": "#ff4092",
                    "stopOpacity": 0.8,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "stop",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "offset": "90%",
                    "stopColor": "white",
                    "stopOpacity": 0,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "stop",
                },
              ],
              "type": "linearGradient",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "",
              "color": "url(#oceanGradient)",
              "data": Array [],
              "getNull": [Function],
              "stack": false,
              "style": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
          null,
          null,
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
